subprojects/base-services/src/main/java/org/gradle/internal/classloader/ClassLoaderUtils.java

```diff
@@ -101,6 +89,63 @@ public static void disableUrlConnectionCaching() {
+        return CLASS_INJECTOR.injectClass(targetClassLoader, className, clazzBytes);
+    }
+
+    private interface ClassInjector {
+        <T> Class<T> injectClass(ClassLoader classLoader, String className, byte[] classBytes);
+    }
+
+    private static class ReflectionClassInjector implements ClassInjector {
+        private final JavaMethod<ClassLoader, Class> defineClassMethod;
+
+        private ReflectionClassInjector() {
+            defineClassMethod = method(ClassLoader.class, Class.class, "defineClass", String.class, byte[].class, int.class, int.class);
+        }
+
+        @Override
+        public <T> Class<T> injectClass(ClassLoader classLoader, String className, byte[] classBytes) {
+            return Cast.uncheckedCast(defineClassMethod.invoke(classLoader, className, classBytes, 0, classBytes.length));
+        }
+    }
+
+    private static class LookupClassInjector implements ClassInjector {
+        private final Class methodHandlesLookupClass;
+        private final JavaMethod methodHandlesLookup;
+        private final JavaMethod methodHandlesPrivateLookupIn;
+        private final JavaMethod lookupFindVirtual;
+        private final MethodType defineClassMethodType;
+
+        private LookupClassInjector() {
+            try {
+                methodHandlesLookupClass = Class.forName("java.lang.invoke.MethodHandles$Lookup");
+            } catch (ClassNotFoundException e) {
+                throw new RuntimeException(e);
+            }
+            methodHandlesLookup = staticMethod(MethodHandles.class, methodHandlesLookupClass, "lookup");
+            methodHandlesPrivateLookupIn = staticMethod(MethodHandles.class, methodHandlesLookupClass, "privateLookupIn", Class.class, methodHandlesLookupClass);
+            lookupFindVirtual = method(methodHandlesLookupClass, MethodHandle.class, "findVirtual", Class.class, String.class, MethodType.class);
+            defineClassMethodType = MethodType.methodType(Class.class, new Class[]{String.class, byte[].class, int.class, int.class});
+        }
+
+        /*
+            This method is equivalent to the following code but use reflection to compile on Java 7:
+
+            MethodHandles.Lookup baseLookup = MethodHandles.lookup();
+            MethodHandles.Lookup lookup = MethodHandles.privateLookupIn(ClassLoader.class, baseLookup);
+            MethodHandle defineClassMethodHandle = lookup.findVirtual(ClassLoader.class, "defineClass", defineClassMethodType);
+            handle.bindTo(classLoader).invokeWithArguments(className, classBytes, 0, classBytes.length));
+         */
+        @Override
+        public <T> Class<T> injectClass(ClassLoader classLoader, String className, byte[] classBytes) {
+            Object baseLookup = methodHandlesLookup.invoke(null);
+            Object lookup = methodHandlesPrivateLookupIn.invoke(null, ClassLoader.class, baseLookup);
```

Comment:
IIUC, I'm not using `Lookup.defineClass`: https://docs.oracle.com/javase/9/docs/api/java/lang/invoke/MethodHandles.Lookup.html#defineClass-byte:A-

This API has the limitation: we must have an existing class within the same package and classloader as the class we're going to define. 

This PR invokes `ClassLoader.defineClass` via `MethodHandles.privateLookupIn` on Java 9+.

Reply:
Using the legacy or new reflection APIs to call the protected ClassLoader.defineClass methods from the wrong context will break once `--illegal-access=deny` is the default. Better to work out how to use Lookup.defineClass.

Reply:
Thanks @AlanBateman for the suggestion. I'll figure this out.

Reply:
@AlanBateman out of curiosity, do you have an idea when this is going to be the default? Because when it happens, it's all Gradle that's going to break because of Groovy. As far as I know Groovy still doesn't have any way to make this warning go away without a serious impact on performance.

Reply:
http://mail.openjdk.java.net/pipermail/jigsaw-dev/2018-March/013688.html

Do you know if Groovy has been tested with `--illegal-access=deny`?

Reply:
Thanks. Groovy breaks with `--illegal-access=deny`. And given that the `CachedClass` model, which is the problematic bit, fails with errors like:
```
WARNING: Illegal reflective access by org.codehaus.groovy.reflection.CachedClass$3$1 (file:/home/cchampeau/.sdkman/candidates/groovy/current/lib/groovy-2.4.7.jar) to method java.lang.Object.finalize()
```
it is no surprise.